fetchYarnBerryDeps: init

yuyuyureka opened the pull request:
## Things done
<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

yuyuyureka committed:
yarn-berry: init fetcher infrastructure

yuyuyureka committed:
hedgedoc: use fetchYarnBerryDeps

yuyuyureka committed:
grafana: use fetchYarnBerryDeps

SuperSandro2000 commented on pkgs/by-name/ya/yarn-berry/fetcher/berry-4-offline.patch:
```diff
@@ -0,0 +1,86 @@
+-            const install = await execUtils.pipevp(`yarn`, [`install`], {cwd, env, stdin, stdout, stderr, end: execUtils.EndStrategy.ErrorCode});
+-            if (install.code !== 0)
+-              return install.code;
++            //const install = await execUtils.pipevp(`yarn`, [`install`], {cwd, env, stdin, stdout, stderr, end: execUtils.EndStrategy.ErrorCode});
```
It is probably easier to maintain for the future, if we delete the code instead of just commenting it.

flokli replied:
Same applies to `berry-3-offline.patch` .

SuperSandro2000 commented:
Noise!

Isn't there a yarn command to add missing hashes or am I imagining things? I would prefer that over maintaining an extra hashes file.

gador reviewed (dismissed):
Just deleting the pgadmin update script (which does not only update the yarn hash) isn't ok. Either ask me how we can migrate it, or try to remove the yarn specific parts. But blandly deleting it is IMHO not ok!
Also I would wish for more details about your deep changes into the yarn ecosystem (preferably also as docs) and an entry into release docs.

gador commented:
> pgadmin4 should be handled by nix-update 22 Apr 2025 08:07:46 Florian ***@***.***>:
> […](#)
> ***@***.**** requested changes on this pull request. Just deleting the pgadmin update script (which does not only update the yarn hash) isn't ok. Either ask me how we can migrate it, or try to remove the yarn specific parts. But blandly deleting it is IMHO not ok! Also I would wish for more details about your deep changes into the yarn ecosystem (preferably also as docs) and an entry into release docs. — Reply to this email directly, view it on GitHub[[#399404 (review)](https://github.com/NixOS/nixpkgs/pull/399404#pullrequestreview-2782857814)], or unsubscribe[https://github.com/notifications/unsubscribe-auth/BOHJVHG5G2SXTRLTPQRST4322XMDBAVCNFSM6AAAAAB3KDFIUCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDOOBSHA2TOOBRGQ]. You are receiving this because you authored the thread. [Tracking image][https://github.com/notifications/beacon/BOHJVHG5FXILKODGHR6LBQ322XMDBA5CNFSM6AAAAAB3KDFIUCWGG33NNVSW45C7OR4XAZNRKB2WY3CSMVYXKZLTORJGK5TJMV32UY3PNVWWK3TUL5UWJTVF34FFM.gif]

I don't believe it does:

<details><Summary>Details</summary>
<p>

```
nix-update pgadmin
$ nix eval --json --impure --expr '
let

  pkgs = import "/home/wogan/software/git/nixpkgs";
  args =  builtins.functionArgs pkgs;
  inputs = (if args ? system then { system = builtins.currentSystem; } else {}) //
           (if args ? overlays then { overlays = [ ]; } else {});
  pkg = (pkgs inputs)."pgadmin";
  sanitizePosition = x: x;

  positionFromMeta = pkg: let
    parts = builtins.match "(.*):([0-9]+)" pkg.meta.position;
  in {
    file = builtins.elemAt parts 0;
    line = builtins.fromJSON (builtins.elemAt parts 1);
  };

  raw_version_position = sanitizePosition (builtins.unsafeGetAttrPos "version" pkg);

  position = if pkg ? isRubyGem then
    raw_version_position
  else if pkg ? isPhpExtension then
    raw_version_position
  else if (builtins.unsafeGetAttrPos "src" pkg) != null then
    sanitizePosition (builtins.unsafeGetAttrPos "src" pkg)
  else
    sanitizePosition (positionFromMeta pkg);
in {
  name = pkg.name;
  old_version = pkg.version or (builtins.parseDrvName pkg.name).version;
  inherit raw_version_position;
  filename = position.file;
  line = position.line;
  urls = pkg.src.urls or null;
  url = pkg.src.url or null;
  rev = pkg.src.rev or null;
  tag = pkg.src.tag or null;
  hash = pkg.src.outputHash or null;
  go_modules = pkg.goModules.outputHash or null;
  go_modules_old = pkg.go-modules.outputHash or null;
  cargo_deps = pkg.cargoDeps.outputHash or null;
  cargo_vendor_deps = pkg.cargoDeps.vendorStaging.outputHash or null;
  raw_cargo_lock =
    if pkg ? cargoDeps.lockFile then
      let
        inherit (pkg.cargoDeps) lockFile;
        res = builtins.tryEval (sanitizePosition {
          file = toString lockFile;
        });
      in
      if res.success then res.value.file else false
    else
      null;
  composer_deps = pkg.composerVendor.outputHash or null;
  composer_deps_old = pkg.composerRepository.outputHash or null;
  npm_deps = pkg.npmDeps.outputHash or null;
  pnpm_deps = pkg.pnpmDeps.outputHash or null;
  yarn_deps = pkg.yarnOfflineCache.outputHash or null;
  yarn_deps_old = pkg.offlineCache.outputHash or null;
  maven_deps = pkg.fetchedMavenDeps.outputHash or null;
  has_nuget_deps = pkg ? nugetDeps;
  mix_deps = pkg.mixFodDeps.outputHash or null;
  tests = builtins.attrNames (pkg.passthru.tests or {});
  has_update_script = pkg.passthru.updateScript or null != null;
  src_homepage = pkg.src.meta.homepage or null;
  changelog = pkg.meta.changelog or null;
  maintainers = pkg.meta.maintainers or null;
}' --extra-experimental-features 'flakes nix-command'
fetch https://github.com/pgadmin-org/pgadmin4/releases.atom
Update REL-9_1 -> REL-9_2 in /home/wogan/software/git/nixpkgs/pkgs/tools/admin/pgadmin/default.nix
$ nix eval --json --impure --expr '
let

  pkgs = import "/home/wogan/software/git/nixpkgs";
  args =  builtins.functionArgs pkgs;
  inputs = (if args ? system then { system = builtins.currentSystem; } else {}) //
           (if args ? overlays then { overlays = [ ]; } else {});
  pkg = (pkgs inputs)."pgadmin";
  sanitizePosition = x: x;

  positionFromMeta = pkg: let
    parts = builtins.match "(.*):([0-9]+)" pkg.meta.position;
  in {
    file = builtins.elemAt parts 0;
    line = builtins.fromJSON (builtins.elemAt parts 1);
  };

  raw_version_position = sanitizePosition (builtins.unsafeGetAttrPos "version" pkg);

  position = if pkg ? isRubyGem then
    raw_version_position
  else if pkg ? isPhpExtension then
    raw_version_position
  else if (builtins.unsafeGetAttrPos "src" pkg) != null then
    sanitizePosition (builtins.unsafeGetAttrPos "src" pkg)
  else
    sanitizePosition (positionFromMeta pkg);
in {
  name = pkg.name;
  old_version = pkg.version or (builtins.parseDrvName pkg.name).version;
  inherit raw_version_position;
  filename = position.file;
  line = position.line;
  urls = pkg.src.urls or null;
  url = pkg.src.url or null;
  rev = pkg.src.rev or null;
  tag = pkg.src.tag or null;
  hash = pkg.src.outputHash or null;
  go_modules = pkg.goModules.outputHash or null;
  go_modules_old = pkg.go-modules.outputHash or null;
  cargo_deps = pkg.cargoDeps.outputHash or null;
  cargo_vendor_deps = pkg.cargoDeps.vendorStaging.outputHash or null;
  raw_cargo_lock =
    if pkg ? cargoDeps.lockFile then
      let
        inherit (pkg.cargoDeps) lockFile;
        res = builtins.tryEval (sanitizePosition {
          file = toString lockFile;
        });
      in
      if res.success then res.value.file else false
    else
      null;
  composer_deps = pkg.composerVendor.outputHash or null;
  composer_deps_old = pkg.composerRepository.outputHash or null;
  npm_deps = pkg.npmDeps.outputHash or null;
  pnpm_deps = pkg.pnpmDeps.outputHash or null;
  yarn_deps = pkg.yarnOfflineCache.outputHash or null;
  yarn_deps_old = pkg.offlineCache.outputHash or null;
  maven_deps = pkg.fetchedMavenDeps.outputHash or null;
  has_nuget_deps = pkg ? nugetDeps;
  mix_deps = pkg.mixFodDeps.outputHash or null;
  tests = builtins.attrNames (pkg.passthru.tests or {});
  has_update_script = pkg.passthru.updateScript or null != null;
  src_homepage = pkg.src.meta.homepage or null;
  changelog = pkg.meta.changelog or null;
  maintainers = pkg.meta.maintainers or null;
}' --extra-experimental-features 'flakes nix-command'
$ nix-build --expr 'let src = (import "/home/wogan/software/git/nixpkgs" (if (builtins.hasAttr "config" (builtins.functionArgs (import "/home/wogan/software/git/nixpkgs"))) then { config.checkMeta = false; overlays = []; } else { }))."pgadmin".src; in (src.overrideAttrs or (f: src // f src)) (_: { outputHash = ""; outputHashAlgo = "sha256"; })' --extra-experimental-features 'flakes nix-command'
warning: found empty hash, assuming 'sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA='
this derivation will be built:
  /nix/store/03lyai2mqby4j9kvwv99wm7ybgp8la1b-source.drv
resolved derivation: '/nix/store/03lyai2mqby4j9kvwv99wm7ybgp8la1b-source.drv' -> '/nix/store/vziv8l2n0dvw8ahc8l12vxr5dk3w39dr-source.drv'...
building '/nix/store/vziv8l2n0dvw8ahc8l12vxr5dk3w39dr-source.drv'...

trying https://github.com/pgadmin-org/pgadmin4/archive/REL-REL_9.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404
Warning: Problem (retrying all errors). Will retry in 1 seconds. 3 retries
Warning: left.
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404
Warning: Problem (retrying all errors). Will retry in 2 seconds. 2 retries
Warning: left.
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404
Warning: Problem (retrying all errors). Will retry in 4 seconds. 1 retries
Warning: left.
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404
error: cannot download source from any mirror
error: builder for '/nix/store/vziv8l2n0dvw8ahc8l12vxr5dk3w39dr-source.drv' failed with exit code 1;
       last 23 log lines:
       >
       > trying https://github.com/pgadmin-org/pgadmin4/archive/REL-REL_9.tar.gz
       >   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
       >                                  Dload  Upload   Total   Spent    Left  Speed
       >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       > curl: (22) The requested URL returned error: 404
       > Warning: Problem (retrying all errors). Will retry in 1 seconds. 3 retries
       > Warning: left.
       >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       > curl: (22) The requested URL returned error: 404
       > Warning: Problem (retrying all errors). Will retry in 2 seconds. 2 retries
       > Warning: left.
       >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       > curl: (22) The requested URL returned error: 404
       > Warning: Problem (retrying all errors). Will retry in 4 seconds. 1 retries
       > Warning: left.
       >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       > curl: (22) The requested URL returned error: 404
       > error: cannot download source from any mirror
       For full logs, run:
         nix log /nix/store/vziv8l2n0dvw8ahc8l12vxr5dk3w39dr-source.drv
error: build of '/nix/store/03lyai2mqby4j9kvwv99wm7ybgp8la1b-source.drv' failed
Traceback (most recent call last):
  File "/nix/store/i5v6x004x6j65a73yin6i5wbiaf8iip7-nix-update-1.11.0/bin/.nix-update-wrapped", line 9, in <module>
    sys.exit(main())
             ^^^^^^
  File "/nix/store/i5v6x004x6j65a73yin6i5wbiaf8iip7-nix-update-1.11.0/lib/python3.12/site-packages/nix_update/__init__.py", line 347, in main
    package = update(options)
              ^^^^^^^^^^^^^^^
  File "/nix/store/i5v6x004x6j65a73yin6i5wbiaf8iip7-nix-update-1.11.0/lib/python3.12/site-packages/nix_update/update.py", line 545, in update
    update_src_hash(opts, package.filename, package.hash)
  File "/nix/store/i5v6x004x6j65a73yin6i5wbiaf8iip7-nix-update-1.11.0/lib/python3.12/site-packages/nix_update/update.py", line 158, in update_src_hash
    target_hash = nix_prefetch(opts, "src")
                  ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/i5v6x004x6j65a73yin6i5wbiaf8iip7-nix-update-1.11.0/lib/python3.12/site-packages/nix_update/update.py", line 143, in nix_prefetch
    raise UpdateError(msg)
nix_update.errors.UpdateError: failed to retrieve hash when trying to update pgadmin.src
```

</p>
</details>

gador commented:
> It _should_ be handled by nix-update, doesn't mean it currently is. I don't know about the specifics of pgadmin, but general support for yarnBerryFetcher in nix-update will be needed for the other packages anyways (nix-update should look at the `offlineCache.berryVersion` attribute to determine which fetcher to use to update the offlineCache hash), 

So in the absence of a working `nix-update` script, how should I (and other maintaners) go about updating the cache?

> and most of what I could see in the pgadmin update script was related to generating the lockfile.

Absolutely. Don't get me wrong. I love to get rid of all that hacky fetching and substituting. So thank you for finding an alternative to #355053 ! I would just have loved some communication first ;-)

flokli committed:
doc: document Yarn v3/v4 infrastructure

Move the existing `yarnConfigHook`, `yarnBuildHook` and
`yarnInstallHook` documentation below a Yarn v1 section.

Also move the documentation for the deprecated `yarn2nix` there, which
is Yarn v1 only.

Co-Authored-By: Yureka <yureka@cyberchaos.dev>

flokli commented:
The docs CI failure only happens when merging into `master`, not with this branch, which is due to https://github.com/NixOS/nixpkgs/pull/400426#issuecomment-2823972916 (revert in https://github.com/NixOS/nixpkgs/pull/401102).

Checking out the branch directly produces working docs.

Ma27 commented on pkgs/by-name/ya/yarn-berry/fetcher/yarn-berry-fetcher.nix:
```diff
@@ -0,0 +1,52 @@
+
+  YARN_ZIP_SUPPORTED_CACHE_VERSION = berryCacheVersion;
+
+  LIBZIP_SYS_USE_PKG_CONFIG = 1;
```
Shouldn't this go into an `env = {}` block?

flokli replied:
Done

Ma27 reviewed (commented):
> The error message could be improved a bit to point to yarn-berry-fetcher missing-hashes.

Either that or point to the manual section you wrote. Not sure how many people would look there otherwise, but you added quite a bunch of useful information in there.

I have a question for my own understanding, but other than that, it looks OK to me (haven't checked the berry fetcher in depth though).

Thanks!

flokli commented:
> [grafana: 11.6.0 -> 11.6.0+security-01](https://github.com/NixOS/nixpkgs/pull/399404/commits/60d24ed4185a29f6c80668641a6b5e974a427a0b) has been added to this PR and should be removed

Yeah, I wanted to pick a parent commit that's before the libblake breakage. With that being reverted in master now I'll rebase in a bit, addressing the comments at the same time.

flokli commented on pkgs/by-name/ya/yarn-berry/fetcher/yarn-berry-fetcher.nix:
```diff
@@ -0,0 +1,52 @@
+    mainProgram = "yarn-berry-fetcher";
+    maintainers = [
+      maintainers.yuka
+      maintainers.flokli
```
If I'd had to guess, to push down the `with` only to when `maintainers` is forced. I applied the suggestion.

gador replied:
@yuyuyureka see #208242 and https://nix.dev/guides/best-practices#with-scopes

gador reviewed (approved):
LGTM. I can't really comment on the patch changes to yarn but the rest looks good to me

gador commented:
@flokli why the rush with the merge? At least you could have waited for @Ma27 to finish the review. Also it may have been a good idea to ask for @doronbehar and @winterqt review on this

flokli commented:
@gador mostly to give this time to stabilize before the stable branchoff. I definitely think we need good berry V3/v4 tooling in the upcoming stable release, the amount of software needing it won't get smaller.

I also had a chat with @winterqt about whether to land this in the Nix/NixOS Matrix channel yesterday, and didn't read @Ma27's review on wanting to do a deeper review and wishing to block until then.

Happy to react on any requested changes / issues that might surface, but I see little reason to block this PR in the meantime.

TomaSajt commented:
I think it would be better if `dontFixup = true` was added to the fetcher by default: fixup phase does the patching of the shebangs, and FODs must never contain nix store paths.

TomaSajt commented:
Another thing that should be addressed IMO:

There are certain node modules that cannot be rebuilt without the network, but have an `install` or `rebuild` script that runs automatically when running `yarn install`.
Some, like `@electron/get` have an env var to disable it, (e.g. `ELECTRON_SKIP_BINARY_DOWNLOAD`,) but other packages might not.

`yarn install` can take the `--mode=skip-build` flag to not run the scripts automatically.

There should be a way to access this behaviour, maybe there should be a `yarnInstallFlags` array.

---

for reference, `npmHooks.npmConfigHook` does something like this:
```bash
npm ci --ignore-scripts $npmInstallFlags $npmFlags
patchShebangs node_modules
npm rebuild $npmRebuildFlags $npmFlags
```
(To solve the problem mentioned above, we could add `--ignore-scripts` to `npmFlags` or `npmRebuildFlags` to make the last line essentially a NOOP)

TomaSajt commented:
> I noticed the issue of patching shebangs before running rebuilds, and was planning something like this (would have been included in the push to your ytmdesktop branch once you responded): [02e3b55](https://github.com/NixOS/nixpkgs/commit/02e3b558b5ffbd754015272ab21f9e4163baf5ad)

You can push to that branch.
I have not encountered anything on that PR that required me to patch the shebangs, so I'm interested to hear where it was needed.

TomaSajt commented:
Okay, another thing I found:

Try using the fetcher on `mx-puppet-discord`'s lockfile.
(do note that it has been marked as broken because it wanted to use the now removed nodejs_18)

The lockfile has some lines with the following resolution value:
```
  resolution: "@mx-puppet/better-discord.js@npm:12.5.1::__archiveUrl=https%3A%2F%2Fgitlab.com%2Fapi%2Fv4%2Fprojects%2F35796068%2Fpackages%2Fnpm%2F%40mx-puppet%2Fbetter-discord.js%2F-%2F%40mx-puppet%2Fbetter-discord.js-12.5.1.tgz"
```

which the fetcher cannot handle

euank commented:
Thanks for the work to get this merged, @yuyuyureka, very much appreciated!
I've been fighting with a lock file over in https://github.com/NixOS/nixpkgs/pull/392737 which this will make way easier.

Unfortunately, one of anki's deps (`license-checker-rseidelsohn`) hits a pathological case in the yarn-fetcher code, so I can't actually get the fetcher to complete.

I've filed an [issue upstream](https://cyberchaos.dev/yuka/yarn-berry-fetcher/-/issues/1), happy to instead file it over here on the `nixpkgs` repo if that seems better

Ma27 commented:
> At least you could have waited for @Ma27 to finish the review.

Sorry, I guess I should've been more explicit here: I don't know much about the current state & internals of JS packaging. I looked at this from the perspective of a package maintainer who takes care of a package (grafana) that's affected by this change.

> node-gyp still needs to run, but this is done in yarnBerryConfigHook automatically by setting npm_config_nodedir and npm_config_node_gyp, (inspired by how npmConfigHook does it)

@yuyuyureka if you don't mind, just a question because I'm not sure I fully understand this: the node-gyp is needed for some dependencies, but not for the things in grafana's missing-hashes file (i.e. @rollup/rollup-..., @parcel/watcher-... etc) since we fetch the native versions here already. And this file is needed because these native things aren't locked because of

> Their reasoning for this is that the local cache files (which are meant to be checked in and distributed over git) for these dependencies are ignored and the platform-specific dependencies are re-fetched on each user's computer

right?

c3n21 commented:
Does this support `pnp`?

I am trying to package verdaccio 6.1.2 which uses yarn berry and pnp for node modules resolution but it seems to break.

You can

[verdaccio.PATCH](https://github.com/user-attachments/files/19964622/verdaccio.PATCH)

TomaSajt commented:
I don't think pnp is supported ATM. Seems like the `packageLocation` fields are all pointing to the yarn cache, but the cache is discarded after the build.

You should somehow configure it to use `node-modules` nodeLinker. Not sure where/how you can do that, but I'll experiment a bit.

If you still want to use the .pnp.cjs method, you'd probably have to copy the cache as well. I have not tried this method yet.

TomaSajt commented:
I made it use node_modules by editing .yarnrc.yml
and then I used npmHooks.npmInstalHook to automatically install the files into $out
I also had to add dontNpmPrune, because it seemed to get stuck at that step. IDK why. Maybe just very very slow.
```nix
  preConfigure = ''
    echo "nodeLinker: node-modules" >> .yarnrc.yml
  '';

  nativeBuildInputs = [
    nodejs
    yarn-berry
    yarn-berry.yarnBerryConfigHook
    npmHooks.npmInstallHook
  ];

  offlineCache = yarn-berry.fetchYarnBerryDeps {
    inherit (finalAttrs) src;
    hash = "sha256-jzkmDxQtIFMa1LIPcvKKsXIItPntgXTavoWhd5eZWyQ=";
  };

  buildPhase = ''
    runHook preBuild
    yarn run build
    runHook postBuild
  '';

  # this would run npm-specific logic that seems to get stuck
  dontNpmPrune = true;
```

`npmHooks.npmInstallHook` is pretty useful for packaging CLIs otherwise installed via `npm install -g`

TomaSajt commented:
After checking 5 random apps that use `fetchYarnBerryDeps` it seems like each one has `nodeLinker: node-modules` in their .yarnrc.yml file.

The current tooling assumes we have node_modules, so I guess we should probably have some code in the hook that makes it use `nodeLinker: node-modules`

Maybe via `yarn config set nodeLinker node-modules`?

TomaSajt commented:
Okay, I came up with this:
```nix
{
  lib,
  stdenv,
  fetchFromGitHub,
  makeWrapper,
  yarn-berry_3,
}:

let
  yarn-berry = yarn-berry_3;
in
stdenv.mkDerivation (finalAttrs: {
  pname = "verdaccio";
  version = "6.1.2";

  src = fetchFromGitHub {
    owner = "verdaccio";
    repo = "verdaccio";
    tag = "v${finalAttrs.version}";
    hash = "sha256-EssvN5HtGI5Hmw4EXetj5nzrkBZAAJGgOx09dlYJzhI=";
  };

  strictDeps = true;

  nativeBuildInputs = [
    makeWrapper
    yarn-berry
  ];

  configurePhase = ''
    runHook preConfigure

    export HOME=$(mktemp -d)
    yarn config set enableTelemetry false

    # all dependencies are vendored in-repo in .yarn/cache
    # we still need to run this because "unplugged packages must be fully materialized on disk to work"
    yarn install

    runHook postConfigure
  '';

  buildPhase = ''
    runHook preBuild
    yarn run build
    runHook postBuild
  '';

  installPhase = ''
    runHook preInstall

    mkdir -p "$out/share"
    cp -r . "$out/share/verdaccio"

    # PnP based pacakges need `yarn node` instead of `node` to run binaries
    makeWrapper ${lib.getExe yarn-berry} "$out/bin/verdaccio" \
      --chdir "$out/share/verdaccio" \
      --add-flags "node ./bin/verdaccio"

    runHook postInstall
  '';
})
```

seems like the main issue was trying to run the binary using `node` instead of `yarn node`, because `node` doesn't respect the .pnp.cjs file.

In this, I just copied everything into `share/verdaccio`, but there may be unnecessary files this way.

TomaSajt commented:
What if we update the config hook so that it detects if you're using PnP via `yarn config get nodeLinker` (it even works if it's unset in the .yarnrc.yml) and it just doesn't replace the local yarn cache + it doesn't run the patchShebangs on the non-existant node_modules directory (assuming we merge that other PR which runs patchShebang node_modules).

So basically it would only do what I did manually in configurePhase.

This code-path wouldn't even require the offlineCache variable.

We could also make it toggleable via a variable instead of using the yarn config, but IMO the config method is cleaner.

SomeoneSerge commented:
Hi! Just to make this explicit, because the manual currently doesn't, are `missing-hashes.json` expected to be non-reproducible/quick-to-rot, and is that the reason we don't wrap them into another FOD?
